server/throttler.py

```diff
@@ -0,0 +1,34 @@
+import asyncio
+from asyncio.locks import Condition
+
+
+class LoginThrottler:
+    def __init__(self, capacity, rate_per_s):
+        self._capacity = capacity
+        self._rate_per_s = rate_per_s
+        self._quota = self._capacity
+        self._wait_lock = Condition()
+        self._leak_coro = asyncio.ensure_future(self._leak())
+        self._closing = False
+
+    async def consume(self):
+        if self._closing:
+            return
+        self._quota -= 1
+        if self._quota < 0:
+            async with self._wait_lock:
+                await self._wait_lock.wait()
+
+    async def _leak(self):
+        while True:
+            self._quota += self._rate_per_s
+            self._quota = min(self._quota, self._capacity)
+            async with self._wait_lock:
+                self._wait_lock.notify(self._rate_per_s)
```

Comment:
Just checking i understand this correctly:
At the start of every second, the quota is increased by `rate_per_s`, and the first `rate_per_s` waiting login requests are released, so there will be `rate_per_s` logins granted at the start of the second and  an additional `rate_per_s` logins granted for new incoming requests during the second?
Then the effective rate would be up to `2 * rate_per_s`?

Intuitively, if you stick with this implementation I would have increased `self._quota` only by `rate_per_s` minus the number of requests being notified.

It's nice that this doesn't need any retries!
I cannot say if it's a drawback to release all of them at once at the start of a second though.

Reply:
The whole thing implements this:
https://en.wikipedia.org/wiki/Leaky_bucket